pkg/modules/modules_linux.go

```diff
@@ -45,15 +63,94 @@ func parseModulesFile(r io.Reader) ([]string, error) {
 	return result, nil
 }
 
-// listModules returns the list of loaded kernel modules names parsed from
-// /proc/modules.
-func listModules() ([]string, error) {
-	fModules, err := os.Open(modulesFilepath)
+// parseBuiltinModulesFile returns the list of builtin kernel modules names.
+func parseBuiltinModulesFile(r io.Reader) ([]string, error) {
+	var result []string
+
+	scanner := bufio.NewScanner(r)
+	scanner.Split(bufio.ScanLines)
+
+	for scanner.Scan() {
+		modulePathRaw := scanner.Text()
+		moduleFileName := filepath.Base(modulePathRaw)
+		moduleFileExt := filepath.Ext(modulePathRaw)
+		moduleName := strings.TrimSuffix(moduleFileName, moduleFileExt)
+		result = append(result, moduleName)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// listLoadedModules returns the parsed list of loaded kernel modules names.
+func listLoadedModules() ([]string, error) {
+	fModules, err := os.Open(loadedModulesFilepath)
 	if err != nil {
 		return nil, fmt.Errorf(
-			"failed to open modules information at %s: %s",
-			modulesFilepath, err)
+			"failed to open loaded modules information at %s: %w",
+			loadedModulesFilepath, err)
 	}
 	defer fModules.Close()
-	return parseModulesFile(fModules)
+	return parseLoadedModulesFile(fModules)
+}
+
+// listBuiltinModules returns the parsed list of builtin kernel modules names.
+func listBuiltinModules() ([]string, error) {
+	var result []string
+
+	locations := []string{
+		"/lib/modules/%s/modules.builtin",
+		"/usr/lib/modules/%s/modules.builtin",
+		"/usr/lib/debug/lib/modules/%s/modules.builtin",
+	}
+
+	release, err := kernelRelease()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, location := range locations {
+		fModulePath := fmt.Sprintf(location, release)
+
+		fModules, err := os.Open(fModulePath)
+		if errors.Is(err, os.ErrNotExist) {
+			continue
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to open builtin modules information at %s: %w",
+				fModulePath, err)
+		}
+
+		defer fModules.Close()
+
+		modules, err := parseBuiltinModulesFile(fModules)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, modules...)
```

Comment:
This will load all available `modules.builtin`, which I think now it should do not and break the execution here, same as in `vendor/github.com/cilium/ebpf/btf/btf.go#findVMLinux`. Am I right on this?

Reply:
I agree, it should use the first one found.

Reply:
Done, also moved the linux-specific tests to the different file. Please resolve this conversation if you are satisfied with the result.